### ISSUE TITLE
Github search sucks

### ISSUE THREAD

**Issue by RenaudWasTaken** [isaacs/github#908]
Hi !

I'm reporting this as I've noticed for quite a while that github search feature is not the
feature needed by a developer.

I often find myself cloning the repository and using grep recursively or on a specific directory.
Advanced search is just plain confusing and takes to many steps to acess.

A few suggestions:
- Enable search on specific directories from the search bar ("This repository" button)
- Replace the search feature with a UI version of grep ?
- Display more results and less code around search term (maybe an expand button)
- Advanced search terms on the search result page 

Thanks!

**Comment by Dorian** [comment]
As an example, searching for "def create" on rails/rails: https://github.com/rails/rails/search?utf8=%E2%9C%93&q=def+create&type=

vs

`git grep "def create"`

<img width="400" alt="screen shot 2017-03-22 at 15 13 33" src="https://cloud.githubusercontent.com/assets/129149/24223163/231e8606-0f12-11e7-8e56-e15c48d01c47.png">

**Comment by FabianKoestring** [comment]
:+1:

**Comment by MaxymVlasov** [comment]
+1

**Comment by RenaudWasTaken** [comment]
By the way you can also upvote this on HN:  https://news.ycombinator.com/item?id=13935370 
This will probably help to get more attention :)

**Comment by davidfurlong** [comment]
Also it seems silly not to be able to search using syntax like (,).; etc - sometimes you need to be able to narrow down results drastically from 1000 to say 10  by using non alphabetic characters

**Comment by hbt** [comment]
@RenaudWasTaken 
what about 

https://github.com/rails/rails/search?utf8=%E2%9C%93&q=filepath%3Aactiverecord+%22def+create%22&type=

specify the directory in `filepath:` and the search term in quotes

**Comment by k1dbl4ck** [comment]
+1

**Comment by Dorian** [comment]
Also, having GitHub search shows tabs with results by default, e.g.

https://github.com/search?q=org%3Arails+some_random_string

<img width="925" alt="screen shot 2017-03-24 at 14 25 51" src="https://cloud.githubusercontent.com/assets/129149/24314234/d5aba830-109d-11e7-966e-518419fb8cd7.png">

https://github.com/search?utf8=%E2%9C%93&q=org%3Arails+belongs_to&type=

<img width="1171" alt="screen shot 2017-03-24 at 14 26 59" src="https://cloud.githubusercontent.com/assets/129149/24314265/fb13b43c-109d-11e7-8933-c58df2184256.png">

**Comment by javadba** [comment]
perfect title : github code search sucks.   I have opened a stackoverflow issue on this.    http://stackoverflow.com/questions/43096264/github-code-search-on-any-branch

**Comment by Mottie** [comment]
There are around 60 search filters available. Even more if you count filter negation; e.g. `-user:Mottie` would exclude me in your searches. So to make it easier, I created a [GitHub Search Autocomplete](https://greasyfork.org/en/scripts/28592-github-search-autocomplete) userscript which makes these filters easier to use and even includes a few examples in a secondary dropdown.

![](https://cloud.githubusercontent.com/assets/136959/24570197/e02272e8-162f-11e7-9ec5-9f70f1820112.gif)

**Comment by leocaseiro** [comment]
I was looking for a way to search a code that starts with, for example, instead of `angular2`, if I would type `angular`, it should find `angularjs`, `angular2`, `angular4` and not just `angular` as it's working atm.

Perhaps this is the right issue.

**Comment by Mottie** [comment]
> Perhaps this is the right issue.

You should use https://github.com/contact to directly contact GitHub with your request. This repo is for us to share and vent our frustrations, I don't think GitHub reads it.

**Comment by Dorian** [comment]
@Mottie Github reads those comments, or at least that's what they said when I met them

**Comment by pakirby1** [comment]
even a basic grep based search would be better.

**Comment by benjamin-rood** [comment]
@pakirby1 It should just be a web interface for `git grep` in the first instance.

**Comment by y2k-shubham** [comment]
Their current implementation is so pathetic that even an integrated Google Custom Search might outpass them

**Comment by andras-ferencz** [comment]
good thing they focused on adding emojis to posts before fixing their broken-by-design search bar

**Comment by birbilis** [comment]
> At most, search results can show two fragments from the same file, but there may be more results within the file. 
https://help.github.com/articles/searching-code/ 

too limited...

**Comment by andybak** [comment]
I'm finding that three quarters of the things I want to use the search for aren't supported.

And local checkout and search doesn't help with Github wide search.

Here's a biggie: 

"Find all repos that contain this string" - pretty basic huh? Instead I get 1000 results for the same repo because that string occurs 1000 times.

I could go on. I find a new one every day.

**Comment by pakirby1** [comment]
Maybe now that they have Microsoft resources they could improve the search.  I’m not holding my breath....

**Comment by komapa** [comment]
Try searching for "?callback=" and see how pathetic this search is...

**Comment by andras-ferencz** [comment]
yeah but remember folks, if you don't have an account and a bunch of free work for random projects that will be dead in a year on github.com you're not a real programmer!

**Comment by birbilis** [comment]
One man's garbage can be another man's treasure, so sharing is caring ;-)
Problem is with a bad search you limit discovery and reuse/adaptations of code.

**Comment by L3nn0x** [comment]
Can we talk about Github search not finding all results? When I do a search I often get less than half the actual results and there's no way to get the other half.

**Comment by Dave9111** [comment]
Yes, its very broken...     When it can't find all of your repositories.... its junk.

**Comment by rajuashok** [comment]
Hello. Has this been looked into? Seriously github search drives me insane. I search for a filename and every mention of that name in other file shows up before that file itself. How is that the right behavior? Also entering the entire path to a file name including the extension (e.g. .tsx) fails completely. Why is including the extension not find the file name?

**Comment by birbilis** [comment]
Maybe it treats the dot specially? Try backslash before the dot. However I guess they only search in the contents

**Comment by honkmother** [comment]
but remember kids, make sure you have a github account on your resume.because nothing says competence and professionalism like using a broken shitty ass website

**Comment by CrackerHax** [comment]
Update: github search still sucks

**Comment by honkmother** [comment]
but omg uber uses github!! for their code!!!so we should use github too!!!

**Comment by ppamorim** [comment]
Searches on my code sometimes returns nothing. Frustrating.

**Comment by IrmantasM** [comment]
Still sucks.

**Comment by TheSharpieOne** [comment]
Here you can see the search is able to determine that 9 results were found for the query within JavasScript files... but when clicking to narrow the search down to see just those 9 JavaScript results... nothing is found (note: the JavaScript results were in the original 59 results... just want to page through them all to find them)

![github-no-results](https://user-images.githubusercontent.com/664714/67305335-2b135c80-f4c3-11e9-98ba-61e261c605cc.gif)

**Comment by pakirby1** [comment]
Search will never be fixed.

**Comment by CrackerHax** [comment]
They might as well link it to google search seeing as to how we end up having to use google to find anything anyway.

**Comment by aderchox** [comment]
GitHub please it's 2020!

Also **the "Most Stars" sort** doesn't work properly at all. Pharo as an instance, has 99% SmallTalk as the programming language of its repository and has 500 stars, but when searching "language:SmallTalk", and choose the "Most Stars" sort type, the repository that is displayed first has around 120 stars only. :|

**Comment by pakirby1** [comment]
They should remove “features” that don’t work correctly.  Unfortunately this doesn’t seem likely.  Also unfortunately but a lot of products offer search functionality that doesn’t work properly.

**Comment by pakirby1** [comment]
https://www.google.com/search?q=why%20does%20search%20suck%20so%20bad&ie=utf-8&oe=utf-8&client=firefox-b-1-m

**Comment by fpesari** [comment]
I was looking for software related to the NES Sound Format (NSF) and I kept getting results about NSFW, even with the double quotes...please fix this, it'd be better if it just ran plain ack or ag in many cases!

**Comment by ipince** [comment]
3.5 years later, it still sucks.

**Comment by pakirby1** [comment]
they should just remove the search capability

**Comment by gittymcgit** [comment]
October 2020 it still sucks

**Comment by shadinaif** [comment]
November 2020, no change

**Comment by fmudhir** [comment]
Still sucks after all this time, can't tell you how much this affects my day-to-day..

**Comment by moe123** [comment]
Still sucks deep, even blows.

**Comment by anko** [comment]
I am having [this mysterious issue](https://github.com/sindresorhus/refined-github/issues/3825#issuecomment-746434999) with search.  In short:  When I'm logged in, I see 0 results for a search that other users see 36 results for.  Other users see results, but I have to log out to see them. :dizzy:

**Comment by TPS** [comment]
I think https://github.blog/changelog/2020-12-17-changes-to-code-search-indexing/ is announcing a step _backwards_.  Thoughts?

**Comment by andybak** [comment]
On Sat, 19 Dec 2020, 16:55 TPS, <notifications@github.com> wrote:

> I think
> https://github.blog/changelog/2020-12-17-changes-to-code-search-indexing/
> is announcing a step *backwards*. Thoughts
>
>
What the goddam? That makes GitHub 50% less useful.

Most of my searches are for code that hasn't changed in the last year.

Someone needs to start an independent Github index.

**Comment by TPS** [comment]
If anyone's interested, please add to the 🔥 by commenting @ https://github.community/t/feedback-on-changes-to-code-search-indexing/150660.  Might not hurt to mention other problems (as above) in passing.

**Comment by jeremiahmoree-adot** [comment]
> I am having [this mysterious issue](https://github.com/sindresorhus/refined-github/issues/3825#issuecomment-746434999) with search. In short: When I'm logged in, I see 0 results for a search that other users see 36 results for. Other users see results, but I have to log out to see them. 💫
> 
> I've emailed Github support about it.
> 
> (Made a comment instead of a separate issue as advised by [#1864 (comment)](https://github.com/isaacs/github/issues/1864#issuecomment-747409641).)

I get 0 results for most searches even when using a string i copied from a file in a repo.  It finds the string in Issues and PRs but not in code which is where I need it to search.  All our work is less than 1 year old so it shouldn't be that problem...

**Comment by mdegans** [comment]
It's hilariously bad. I have a repo named `libptzf` and it can't find it in `Youre Repositories` searching for `ptzf`. Did i put a dash after the lib or not? I can never remember. Search should be able to help.

**Comment by Levi-Lesches** [comment]
Reminder that this is not [the official GitHub feedback repo](https://github.com/github/feedback/discussions/categories/general-feedback). Open a discussion there to get GItHub's attention.

**Comment by pakirby1** [comment]
Maybe it’s time to email Nat Friedman (GitHub CEO) this thread

**Comment by Levi-Lesches** [comment]
> Reminder that this is not [the official GitHub feedback repo](https://github.com/github/feedback/discussions/categories/general-feedback). Open a discussion there to get GItHub's attention.

You don't have to email anyone, just open a new discussion in the official repo above instead of here. GitHub is offering you a way to interact with their staff, which wasn't possible like this before, why ignore it? This repo was made because there wasn't such an option at the time. Now there is.

**Comment by pakirby1** [comment]
https://github.com/github/feedback/discussions/4436

**Comment by CrackerHax** [comment]
search is just completely unusable now.
